// @mui/internal-test-utils/build/fireDiscreteEvent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mouseUp = exports.mouseDown = exports.keyUp = exports.keyDown = exports.click = void 0;
const React = __importStar(require("react"));
const react_1 = require("@testing-library/react");
const noWrapper = (callback) => callback();
/**
 * @param {() => void} callback
 * @returns {void}
 */
function withMissingActWarningsIgnored(callback) {
    if (React.version.startsWith('18')) {
        callback();
        return;
    }
    const originalConsoleError = console.error;
    console.error = function silenceMissingActWarnings(message, ...args) {
        const isMissingActWarning = /not wrapped in act\(...\)/.test(message);
        if (!isMissingActWarning) {
            originalConsoleError.call(console, message, ...args);
        }
    };
    const originalConfig = (0, react_1.getConfig)();
    (0, react_1.configure)({
        eventWrapper: noWrapper,
    });
    try {
        callback();
    }
    finally {
        (0, react_1.configure)(originalConfig);
        console.error = originalConsoleError;
    }
}
// -----------------------------------------
// WARNING ⚠️ WARNING ⚠️ WARNING ⚠️ WARNING
//
// Do not add events here because you want to ignore "missing act()" warnings.
// Only add events if you made sure that React actually considers these as "discrete".
// Be aware that "discrete events" are an implementation detail of React.
// To test discrete events we cannot use `fireEvent` from `@testing-library/react` because they are all wrapped in `act`.
// `act` overrides the "discrete event" semantics with "batching" semantics: https://github.com/facebook/react/blob/3fbd47b86285b6b7bdeab66d29c85951a84d4525/packages/react-reconciler/src/ReactFiberWorkLoop.old.js#L1061-L1064
// Note that using `fireEvent` from `@testing-library/dom` would not work since /react configures both `fireEvent` to use `act` as a wrapper.
// -----------------------------------------
function click(...args) {
    return withMissingActWarningsIgnored(() => {
        react_1.fireEvent.click(...args);
    });
}
exports.click = click;
function keyDown(...args) {
    return withMissingActWarningsIgnored(() => {
        react_1.fireEvent.keyDown(...args);
    });
}
exports.keyDown = keyDown;
function keyUp(...args) {
    return withMissingActWarningsIgnored(() => {
        react_1.fireEvent.keyUp(...args);
    });
}
exports.keyUp = keyUp;
function mouseDown(...args) {
    return withMissingActWarningsIgnored(() => {
        react_1.fireEvent.mouseDown(...args);
    });
}
exports.mouseDown = mouseDown;
function mouseUp(...args) {
    return withMissingActWarningsIgnored(() => {
        react_1.fireEvent.mouseUp(...args);
    });
}
exports.mouseUp = mouseUp;
