// @mui/internal-test-utils/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.strictModeDoubleLoggingSuppressed = exports.flushMicrotasks = exports.userEvent = exports.fireDiscreteEvent = exports.programmaticFocusTriggersFocusVisible = exports.simulatePointerDevice = exports.focusVisible = exports.wrapsIntrinsicElement = exports.findOutermostIntrinsic = exports.createMount = exports.createDescribe = exports.describeConformance = void 0;
const React = __importStar(require("react"));
__exportStar(require("./components"), exports);
var describeConformance_1 = require("./describeConformance");
Object.defineProperty(exports, "describeConformance", { enumerable: true, get: function () { return __importDefault(describeConformance_1).default; } });
__exportStar(require("./describeConformance"), exports);
var createDescribe_1 = require("./createDescribe");
Object.defineProperty(exports, "createDescribe", { enumerable: true, get: function () { return __importDefault(createDescribe_1).default; } });
__exportStar(require("./createRenderer"), exports);
var createMount_1 = require("./createMount");
Object.defineProperty(exports, "createMount", { enumerable: true, get: function () { return __importDefault(createMount_1).default; } });
var findOutermostIntrinsic_1 = require("./findOutermostIntrinsic");
Object.defineProperty(exports, "findOutermostIntrinsic", { enumerable: true, get: function () { return __importDefault(findOutermostIntrinsic_1).default; } });
Object.defineProperty(exports, "wrapsIntrinsicElement", { enumerable: true, get: function () { return findOutermostIntrinsic_1.wrapsIntrinsicElement; } });
var focusVisible_1 = require("./focusVisible");
Object.defineProperty(exports, "focusVisible", { enumerable: true, get: function () { return __importDefault(focusVisible_1).default; } });
Object.defineProperty(exports, "simulatePointerDevice", { enumerable: true, get: function () { return focusVisible_1.simulatePointerDevice; } });
Object.defineProperty(exports, "programmaticFocusTriggersFocusVisible", { enumerable: true, get: function () { return focusVisible_1.programmaticFocusTriggersFocusVisible; } });
exports.fireDiscreteEvent = __importStar(require("./fireDiscreteEvent"));
exports.userEvent = __importStar(require("./userEvent"));
var flushMicrotasks_1 = require("./flushMicrotasks");
Object.defineProperty(exports, "flushMicrotasks", { enumerable: true, get: function () { return __importDefault(flushMicrotasks_1).default; } });
/**
 * Set to true if console logs during [lifecycles that are invoked twice in `React.StrictMode`](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects) are suppressed.
 * Useful for asserting on `console.warn` or `console.error` via `toErrorDev()`.
 */
exports.strictModeDoubleLoggingSuppressed = React.version.startsWith('17');
